joomla/joomla-cms
# Fix inconsisteny between new install and update for NOT NULL of urls columns of redirect_links table with MySQL or Microsoft databases

Pull Request for new issue.

Only relevant for MySQL or Microsoft SQL Server or SQL Azure databases, i.e. **not** relevant for PostgreSQL databases.
#### Summary of Changes

With PR #9269 for enlargement of URL colums in the "redirect_links" table, a "NOT NULL" constraint for the "new_url" column was accidently added in schema update scripts "3.5.2-2016-04-06.sql" for mysql and sqlazure, which does not exist in the create table statement for new installation.

This PR corrects these inconsistencies between update and new install by new schema update scripts for updating from a 3.5.0 or 3.5.1.

In addition, the old schema updates "3.5.2-2016-04-06.sql" are corrected, too, so pre-3.5.0 versions will not be updated in the wrong way.

@shur Thanks for reporting this issue to me.
#### Testing Instructions
##### Pre-requisites, important!!!
1. This PR cannot be tested on a newly installed current staging, because there the database schema is of course consistent, so the issue does not appear.
2. Due to the limited capabilities of the database schema manager aka "database fix", you cannot use patchtester and then do a database fix to apply this PR.
3. For the same reason, you cannot use the "copy or unzip the files on the server and then run the database fix" method for updating to latest staging + this PR, you have to use the custom URL and Joomla! Update Component.

So as a starting point, use a pre-3.5.0, e.g. a 3.4.8 for this test. No need to test several older versions, a 3.4.8 is sufficient.

Finally, please mention in the comment which database type(s) you have tested when setting your test result.
##### Test Steps

Step 1: On a Joomla! pre-3.5.0, e.g. a 3.4.8, check e.g. with phpMyAdmin (or in case of Microsoft stuff, the Microsoft tool) that in your database, column "new_url" of the redirect_links table does not have a "NOT NULL" constraint, i.e. NULL values are allowed, and columns "old_url" and "referer" have a "NOT NULL" constraint, i.e. NULL values are not allowed.

Result: "NOT NULL" constraints are as described.

Step 2: Update to 3.5.0 or 3.5.1 (does not matter which one).

Step 3: Check e.g. e.g. with phpMyAdmin (MS tool) the "NOT NULL" constrains of columns "old_url", "new_url" and "referer" of the redirect_links table in your database.

Result: There is now a "NOT NULL" constraint for column "new_url". The other 2 columns have not changed.

Step 4: Check with a text editor or on command line (e.g. on Linux: less command) the "NOT NULL" constrains of the "new_url" column in the SQL statement where the table is created in file "installation/sql/<dbtype>/joomla.sql", where <dbtype> is your database server type mysql, postgresql or sqlazure (= sqlsrv).

Result: According to the joomla.sql the "NOT NULL" constraint shouldn't have changed compared to Step 1.

Step 5: Update to current staging + this PR = a patched 3.6.0-dev, using following custom URL: [http://test5.richard-fath.de/list_test7.xml](http://test5.richard-fath.de/list_test7.xml).

Step 6: Check e.g.with phpMyAdmin (MS tool) the "NOT NULL" constrains of columns "old_url", "new_url" and "referer" of the redirect_links table in your database.

Result: The "NOT NULL" constrains of columns "old_url", "new_url" and "referer" of the redirect_links table are as described in Step 1. 

Step 7: Check with a text editor or on command line (e.g. on Linux: less command) the "NOT NULL" constrains of these columns in the SQL statement where the table is created in file "installation/sql/<dbtype>/joomla.sql"

Result: Also according to the joomla.sql the "NOT NULL" constraints shouldn't have changed compared to Step 1.
